helix-rest/src/main/java/org/apache/helix/rest/server/resources/mock/MockMetadataStoreDirectoryAccessor.java

```diff
@@ -0,0 +1,118 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import javax.ws.rs.Path;
+
+import org.apache.helix.msdcommon.datamodel.TrieRoutingData;
+import org.apache.helix.msdcommon.exception.InvalidRoutingDataException;
+import org.apache.helix.rest.metadatastore.MetadataStoreDirectory;
+import org.apache.helix.rest.metadatastore.ZkMetadataStoreDirectory;
+import org.apache.helix.rest.metadatastore.accessor.MetadataStoreRoutingDataReader;
+import org.apache.helix.rest.metadatastore.accessor.MetadataStoreRoutingDataWriter;
+import org.apache.helix.rest.metadatastore.accessor.ZkRoutingDataReader;
+import org.apache.helix.rest.metadatastore.accessor.ZkRoutingDataWriter;
+import org.apache.helix.rest.server.resources.metadatastore.MetadataStoreDirectoryAccessor;
+
+
+/**
+ * An accessor that mocks the MetadataStoreDirectoryAccessor for testing purpose.
+ */
+@Path("/mock")
+public class MockMetadataStoreDirectoryAccessor extends MetadataStoreDirectoryAccessor {
+  // A flag that will be modified if the underlying MockZkRoutingDataWriter makes an operation
+  // against ZooKeeper
+  public static boolean operatedOnZk = false;
+  // The instance of mockMSD that's created by this accessor; it's saved here to be closed later
+  public static MetadataStoreDirectory _mockMSDInstance;
+
+  /**
+   * This method is overriden so that an instance of MockZkMetadataStoreDirectory can be passed in
+   */
+  @Override
+  protected void buildMetadataStoreDirectory(String namespace, String address) {
+    try {
+      _metadataStoreDirectory = new MockZkMetadataStoreDirectory(namespace, address);
+      _mockMSDInstance = _metadataStoreDirectory;
+    } catch (InvalidRoutingDataException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Used to artificially create another instance of ZkMetadataStoreDirectory.
+   * ZkMetadataStoreDirectory being a singleton makes it impossible to test it,
+   * therefore this is the only way to create another instance.
+   */
+  class MockZkMetadataStoreDirectory extends ZkMetadataStoreDirectory {
+    MockZkMetadataStoreDirectory(String namespace, String zkAddress)
+        throws InvalidRoutingDataException {
+      super();
+
+      // Manually populate the map so that MockZkRoutingDataWriter can be passed in
+      _routingZkAddressMap.put(namespace, zkAddress);
+      _routingDataReaderMap.put(namespace, new ZkRoutingDataReader(namespace, zkAddress, this));
+      _routingDataWriterMap.put(namespace, new MockZkRoutingDataWriter(namespace, zkAddress));
+      _realmToShardingKeysMap.put(namespace, _routingDataReaderMap.get(namespace).getRoutingData());
+      _routingDataMap.put(namespace, new TrieRoutingData(_realmToShardingKeysMap.get(namespace)));
+    }
+
+    @Override
+    public void close() {
+      _routingDataReaderMap.values().forEach(MetadataStoreRoutingDataReader::close);
+      _routingDataWriterMap.values().forEach(MetadataStoreRoutingDataWriter::close);
```

Comment:
Nit: Should you clear the maps?

Reply:
It's not necessary because MockMSD is not a singleton. A new instance is created each time.

Reply:
OK.